keps/draft-20180418-security-profile.md

```diff
@@ -0,0 +1,1516 @@
+without them having to understand the low-level configuration used to achieve it.
+- The profile is sharable and reusable as it’s defined in a portable way
+- Enforced automatically and continuously
+
+Additionally, a Security Profile
+
+- Is versioned, to allow seamless upgrade of the cluster and Security Profile
+- Is highly customizable for particular needs
+- Is highly extendable with extra features
+
+An example if using security profile is
+
+```
+kubeadm init --security-profile=default-1.0-0
+```
+
+The command above will configure a secure cluster according to security profile `default-1.0-0`.
+
+As security profiles cover common/generic use cases,
+they are portable and reusable in most Kubernetes deployments,
+They are **NOT** defining policies specific to a particular Kubernetes deployment, like:
+
+- ResourceQuota
+- RBAC rules
+- Rules in NetworkPolicy for particular workloads
+
+The proposal also works with Config-as-Code systems,
+see the [section](#work-with-config-as-code-systems) below for details.
+
+### Proposed Security Profiles
+
+#### default
+
+Full example: [default-1.0-0](#the-default-profile)
+
+Minimum security practice to be compatible with most workloads.
+Best fit for sharing the cluster with trusted workloads.
+
+Enforcement:
+
+- etcd
+  - TLS required for communication
+  - Explicitly specify `--wal-dir`, and `--max-wals=0`
+- kube-apiserver
+  - TLS required for communication
+  - Audit logging must be explicitly configured
+  - No anonymous auth, basic auth (unset `--anonymous-auth`, `--basic-auth-file`)
+  - Required authorization-mode: Node, RBAC
+  - Excluded authorization-mode: ABAC, AlwaysAllow
+  - Disable insecure port (`--insecure-port=0`, unset `--insecure-bind-address`)
+  - No auth token file (unset `--token-auth-file`)
+  - Disable profiling (unset `--profiling`)
+  - Request timeout is explicitly set
+  - Use service account credentials (`--use-service-account-credentials=true`)
+  - Required admission controls: AlwaysPullImages, DenyEscalatingExec, NamespaceLifecycle, ResourceQuota, PodSecurityPolicy, ServiceAccount, NodeRestriction, EventRateLimit
+  - Excluded admission controls: AlwaysAdmit
+  - Required feature gates to be ON: AdvancedAuditing, RotateKubeletServerCertificate, RotateKubeletClientCertificate
+- controller-manager
+  - Explicitly set `--terminated-pod-gc-threshold`
+- kubelet
+  - Disable hostname override (unset `--hostname-override`)
+  - No privileged
+- addons
+  - No dashboard
+- file-permissions
+  - ... # correct setup of cert/key/config file permissions
+- PodSecurityPolicy
+  - No priviledged
+- NodeTaints
+  - No schedule on master
+
+#### saas-multitenancy
+
+Full Example: [saas-multitenancy-1.0-0](#the-saas-multitenancy-profile)
+
+For SaaS scenarios which allows only plain workloads without the need to gain privileged access.
+
+Enforcement: in addition to `default`
+
+- PodSecurityPolicy
+  - No privileged pods
```

Comment:
Isn't this already in 'default'?

Reply:
I should remove that from `default`.

Reply:
so 'default' would allow privileged pods?
Obviously I'm coming at this from a security angle, but if I turn on security profiles, I'm probably OK with it warning me that I shouldn't be running pods as privileged and then fix that - or are you saying that there isn't a lower level so they're stuck with that issue?

Reply:
This is my personal opinion (I'd like to hear other thoughts, as I'm probably wrong): the `default` security profile is provided to bring in the lowest level (there's no lower levels) of security (eliminate some really bad insecure configurations) without breaking compatibility of how people are using Kubernetes today. So we can use `default` to help Kubernetes users smoothly transit from insecure to secure. I believe privileged containers are allowed in most Kubernetes clusters with trusted users (like inside the same corp or team) and sometimes useful. That's the reason I removed "no-privileged containers" from `default` profile.

Reply:
got it, I see you defined default as "to be compatible with most workloads". That makes sense then.